### ISSUE TITLE
[HOLD for payment 2023-09-27] [$500] Web - The delete confirmation modal fails to close when removing a member from another tab

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#26409]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Navigate to Settings > Workspace > New workspace.
2. Go to the Members section.
3. Invite a member to the workspace.
4. Select the member and click on "Remove."
5. Copy the URL and paste it into another tab.
6. In the new tab, remove the member.
7. Return to the previous tab.
8. The delete confirmation modal does not close, and the member is removed.

## Expected Result:
The delete confirmation modal should close when a member is removed from another tab

## Actual Result:
 The delete confirmation modal does not close when removing a member from another tab

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.60.1
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/93399543/6b406423-6640-4fe5-8094-6922509bfa13

https://github.com/Expensify/App/assets/93399543/3fefc617-f6a7-4402-a959-7b05bad9a773

**Expensify/Expensify Issue URL:**
**Issue reported by:** @ayazhussain79
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1692912070414399

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01f4dfb3829fccd609</li>
        <li>Upwork Job ID: 1699049489764618240</li>
        <li>Last Price Increase: 2023-09-12</li>
<li>Automatic offers: </li>
<ul>
<li>mollfpr | Reviewer | 26664911</li>
<li>DylanDylann | Contributor | 26664913</li>
</ul></li>
    </ul>
</details>

**Comment by bfitzexpensify** [comment]
Sending this external

**Comment by mollfpr** [comment]
I think the modal should be dismissed whenever the list member is changed (added or removed member).

@bfitzexpensify What do you think?

**Comment by bfitzexpensify** [comment]
I agree, that behaviour makes sense to me

**Comment by DylanDylann** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Web - The delete confirmation modal fails to close when removing a member from another tab

### What is the root cause of that problem?
We don't call setRemoveMembersConfirmModalVisible(false) when the number of members changes

### What changes do you think we should make in order to solve the problem?
We can dismiss the modal when the number of members change. So that, in case the information of the user is updated we will not dismiss modal
Note that: we are using `data` variables to display member list. So we should use data.length to check if the number of members changes and dismiss modal

```
useEffect(() => {
        setRemoveMembersConfirmModalVisible(false);
    }, [data.length]);

```

**Comment by mollfpr** [comment]
>In my opinion, if any member of a workspace is removed and the modal is open, it should close so that the user can re consider. To implement this, we need to add setRemoveMembersConfirmModalVisible(false); here:


@GItGudRatio Could you be more specific, is it inside the `useEffect` or the `useState` callback?

---

@DylanDylann May I ask why we're not using the `props.policyMembers` instead for the dependencies?

**Comment by bfitzexpensify** [comment]
@mollfpr how are things looking here? Seems like we're still working on finding the right proposal?

**Comment by mollfpr** [comment]
Sorry for the delay! Testing the proposals now.

**Comment by mollfpr** [comment]
>
@mollfpr we can use props.personaldetail.length or data.length as dependency but i think we shouldn't use personalDetail or data because we only want dismiss modal if the number of member change

@DylanDylann I'm not sure I understand your point here. Your proposal is using `data.length`.

@GItGudRatio Could you update your RCA for this expected behavior [here](https://github.com/Expensify/App/issues/26409#issuecomment-1708508577)? Your solution seems to work for removed users and when new users are added. Just to clarify that you're RCA is correct with the new expected.

**Comment by DylanDylann** [comment]
@mollfpr We don't want to dismiss the modal if the display name of a member changes, right? So that, we only should put data.length or props.policyMembers.length as dependency of useEfffect

**Comment by DylanDylann** [comment]
Updated proposal: https://github.com/Expensify/App/issues/26409#issuecomment-1709815290 with alternative solution

**Comment by mollfpr** [comment]
> We don't want to dismiss the modal if the display name of a member changes, right?

@DylanDylann Do you mean we can't just use `props.policyMembers` as dependencies because it will close the modal on any user that changes their display name?

I just tested the above case, and it doesn't dismiss the modal when other members change their display name.

**Comment by DylanDylann** [comment]
> We don't want to dismiss the modal if the display name of a member changes, right?

Sorry about this. But let's see policyMembers

<img width="402" alt="Screenshot 2023-09-12 at 00 11 04" src="https://github.com/Expensify/App/assets/141406735/49b1db19-f810-4fb4-abb6-394a8f2f2320">

If we put props.policyMembers as dependencies, when the errors field or pendingAction field is changed we also dismiss modal

**Comment by DylanDylann** [comment]
@mollfpr You could try with this step:
Scenario 1:
1. Open workspace member page
2. Go offline
3. Invite new member 
4. Click on another member and clicking remove --> the Remove members modal will appear
5. Go online

Verify that:
1. Using props.policyMembers as dependency:  the Remove members modal will disappear after step 5
2. Using _.size(props.policyMembers) as dependency:  the Remove members modal will remain after step 5

**Comment by mollfpr** [comment]
>Using _.size(props.policyMembers) as dependency: the Remove members modal will remain after step 5

@DylanDylann Why is the modal not dismissed? I think we are already clear [here](https://github.com/Expensify/App/issues/26409#issuecomment-1708508577) that the modal will dismiss the members changed, either because of removing or adding a new member.

**Comment by DylanDylann** [comment]
Agree that the modal will dismiss the members change. But in the above case, we don't add new member we only update pendingAction Field of that member

**Comment by mollfpr** [comment]
>But in the above case, we don't add new member we only update pendingAction Field of that member

@DylanDylann Could you clarify this? After it goes online the other tab with the modal open will receive the pusher notification and update the policy members accordingly.

**Comment by mollfpr** [comment]
@DylanDylann I think it's better if you could attach a video for that, so I can see maybe I'm doing it wrong.

**Comment by mollfpr** [comment]
@DylanDylann Using the `_.size(props.policyMembers)` also has another issue.


https://github.com/Expensify/App/assets/25520267/cad10224-90bf-4728-98b2-cd8e96fa3b41

As you can see at the end, the modal is not dismissed, but the selected member is unselected. It will be confusing which member will be removed.

**Comment by DylanDylann** [comment]
@mollfpr This bug only happens if we invite a new user (not exist yet). When we invite a new user (not exist yet) we will create a new optimistic accountID and then when coming back online the BE will return another accountID for that user (BE don't return the optimistic accountID). So that the number of members is not changed  

So that, I have another suggest, we can use usePrevious to save the accountIDs and dismiss modal when the accountIDs changed like this
```
const accountIDs = useMemo(() => _.keys(props.policyMembers), [props.policyMembers])
    const prevAccountIDs = usePrevious(accountIDs);
    useEffect(() => {
        if (removeMembersConfirmModalVisible && !_.isEqual(accountIDs, prevAccountIDs)) {
            setRemoveMembersConfirmModalVisible(false);
        }    
        setSelectedEmployees((prevSelected) =>
            _.intersection(
                prevSelected,
                _.map(_.values(PolicyUtils.getMemberAccountIDsForWorkspace(props.policyMembers, props.personalDetails)), (accountID) => Number(accountID)),
            ),
        );
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [props.policyMembers]);
```

I tested and It worked well for all above cases

**Comment by mollfpr** [comment]
@DylanDylann What case should we compare the accountIDs? It seems that the confirmation modal will be dismiss everytime on the members changed.

**Comment by mollfpr** [comment]
So no different with just dismissing depending on the `props.policyMembers`?

**Comment by DylanDylann** [comment]
@mollfpr 
Let's see scenerio 1 in [this comment](https://github.com/Expensify/App/issues/26409#issuecomment-1715594941)
> Using props.policyMembers as dependency: the Remove members modal will disappear after step 5

**Comment by mollfpr** [comment]
>Scenerio 1 in this comment: the Remove members modal will remain after step 5

I tried this, but it dismissed the modal.

**Comment by mollfpr** [comment]
Let me clarify it:

1. Adding new user _will_ dismiss the modal
2. Adding existing user _will not_ dismiss the modal
3. Remove any member _will_ dismiss the modal

@DylanDylann Is this correct?

**Comment by DylanDylann** [comment]
Agree that. in the point 1, we should dismiss modal because when invite new user offline (not exist yet) we will create optimistic accountID and when coming back online the BE will return another accountID for that user (you also see that the avatar of that user changed).

**Comment by mollfpr** [comment]
Thanks @DylanDylann, that looks good. Could you update your proposal with your latest solution? Also, put the above analysis to your RCA too.

**Comment by mollfpr** [comment]
Thank you @DylanDylann, for bearing til the end and coming up with the proper solution.

We have agreed on the expectation [here](https://github.com/Expensify/App/issues/26409#issuecomment-1708508577), that the modal will be dismissed whenever the members list changes.

We also find the proper action to dismiss the modal defined by the situation on:

1. Adding a new user _**will**_ dismiss the modal
2. Adding existing user **_will not_** dismiss the modal
3. Remove any member **_will_** dismiss the modal

The [proposal](https://github.com/Expensify/App/issues/26409#issuecomment-1709815290) from @DylanDylann handles the above case well.

🎀 👀 🎀 C+ reviewed!

**Comment by DylanDylann** [comment]
@mollfpr The PR is ready for review

**Comment by mollfpr** [comment]
> [@mollfpr] The PR that introduced the bug has been identified. Link to the PR:
 [@mollfpr] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:

No offending PRs. This improvement is to close the modal whenever the list is changed.

> [@mollfpr] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:

The regression step should be enough.

>[@mollfpr] Determine if we should create a regression test for this bug.
 [@mollfpr] If we decide to create a [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) for the bug, please propose the regression test steps to ensure the same bug will not reach production again.

1. Navigate to Settings > Workspace > New workspace.
2. Go to the Members section.

**The confirmation modal should be closed when a new member is added in another tab/device**

3. Invite a member to the workspace.
4. Select the member and click on "Remove."
5. Copy the URL and paste it into another tab.
6. In the new tab, remove the member.
7. Return to the previous tab.
8. Verify that the delete confirmation modal is closed

**The confirmation modal should be closed when a new user is added from offline**

3. Go offline
4. Invite a member (not exist yet like "[qweqwe@gmail.com](mailto:qweqwe@gmail.com)") to the workspace.
5. Select the member and click on "Remove."
6. Go online
7. Verify that the delete confirmation modal is closed

**The confirmation modal remains when existed user is added from offline**

3. Go offline
4. Invite a member (who we chatted before) to the workspace.
5. Select the member and click on "Remove."
6. Go online
7. Verify that the delete confirmation modal still remained

**Comment by bfitzexpensify** [comment]
All payments complete, issue opened for the regression test, we're good to close this one out. Thanks for the work, everyone!
